### ISSUE TITLE
`bearriver`: add `dpSwitch`

### ISSUE THREAD

**Issue by geroldmeisinger** [ivanperez-keera/dunai#331]
BearRiver is missing some definitions available in Yampa. One such definition is `dpSwitch`.

an implementation of `dpSwitch` has been provided by Jonathan Thaler at https://github.com/ivanperez-keera/dunai/issues/40#issuecomment-354439225:
```haskell
dpSwitch :: (Monad m, Traversable col)
         => (forall sf. (a -> col sf -> col (b, sf)))
         -> col (SF m b c) 
         -> SF m (a, col c) (Event d) 
         -> (col (SF m b c) -> d -> SF m a (col c))
         -> SF m a (col c)
dpSwitch rf sfs sfF sfCs = MSF $ \a -> do
  let bsfs = rf a sfs
  res <- T.mapM (\(b, sf) -> unMSF sf b) bsfs
  let cs   = fmap fst res
      sfs' = fmap snd res
  (e,sfF') <- unMSF sfF (a, cs)
  let ct = case e of
          Event d -> sfCs sfs' d
          NoEvent -> dpSwitch rf sfs' sfF' sfCs
  return (cs, ct)
```

EDIT: Ivan Perez: Added a short description and brought in text from different issue to help process this independently without jumping around.

**Comment by ivanperez-keera** [comment]
The steps required to determine if we can add this implementation are:
- Is it extensionally the same as the one in Yampa? One can test them by hand, but https://github.com/ivanperez-keera/dunai/commit/9285e4edf22cec1176c7e881c64d38035f771802 offers a systematic way of checking that an implementation in Bearriver and one in Yampa behave the same.
- Does it leak? For that it would be good to run it with a long trace and see that performance does not degrade. You can use the same approach as in step (1) to generate very long traces, but make sure you `whnf` the outputs.
- Could it be simplified? In particular, is there a way to write it using existing combinators without making the code harder to understand?
- Is the code understandable? From re-writing to renaming variables to adding comments; if it's not very obvious to you how it works, it won't be understandable to others. Avoid repetition of work by helping others understand it. (Hint: it isn't)
- Is it conformant with our style guide? See https://github.com/ivanperez-keera/dunai#style for details. (Hint: it isn't)

**Comment by ivanperez-keera** [comment]
For context: I've been adding missing definitions as subissues of #40. If you look at the [recent history](https://github.com/ivanperez-keera/dunai/issues/40#ref-issue-1680417008) of that issue, you'll see that several modules from Yampa have been added to BearRiver. The module `Switches` now exists, but it's incomplete. The proposed `dpSwitch` would fit in that module, but it'll take me some time to get there: I'm trying to go in order so that I don't miss things.

**Comment by ivanperez-keera** [comment]
Closed in favor of #425 .
